Source/WebCore/loader/FrameMemoryMonitor.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "FrameMemoryMonitor.h"
+
+#include "Document.h"
+#include "FrameLoader.h"
+#include "LocalFrameLoaderClient.h"
+
+namespace WebCore {
+
+Ref<FrameMemoryMonitor> FrameMemoryMonitor::create(LocalFrame& frame)
+{
+    return adoptRef(*new FrameMemoryMonitor(frame));
+}
+
+FrameMemoryMonitor::FrameMemoryMonitor(LocalFrame& frame)
+    : m_frame(frame)
+{
+}
+
+FrameMemoryMonitor::~FrameMemoryMonitor()
+{
+}
+
+void FrameMemoryMonitor::addUsage(size_t bytes)
+{
+    if (m_usageHasExceeded)
+        return;
+
+    // FIXME: Add support for memory usage deltas
+    m_currentMemoryUsage = bytes;
```

Comment:
This method name seems confusing -- It's really "setUsage" for now. Maybe we should use `setUsage` as the name until we have some mechanism for deltas?

Reply:
hmm... the current logic can underreport sometimes and we report memory upwards, so it should always accumulate, which means we need HTMLMediaElement to manage these deltas?

since we're reporting process-level memory (rather than per-object), does the last frame win here? assuming there is an idle period between the frames below or memory is being freed between reports

Root Frame (FrameMemoryMonitor)
- Frame A (100 MB)
- Frame B (50 MB)

Reply:
Does that mean that each `addMemory` message represents the current memory use of the media element? The way I thought this would work was that each media element uses some amount of memory. When we detect that one of the elements is using exceptional amounts of memory (perhaps for more than X measurements or something) that we would want to take action.

I'm not sure how holding a single value looking at all frames would be the right design for that. Shouldn't we be tracking it in some kind of table of "media element" to "memory use" or something?

Reply:
`addUsage` (or `setUsage`) reflects the current memory snapshot from the GPUP. we're invoking it in a media player's tick, implicitly tying the memory usage to a specific media element.

the current limitation i'm concerned about is there's no attribution to which subframe caused the spike. so, when a threshold is exceeded at the root frame, we don't know at which depth the frame is responsible. iirc, there's a limit of 32 depth of frames.

the core issue is this:

```c++
if (RefPtr parentMonitor = parentFrameMemoryMonitorIfExists())
        parentMonitor->setUsage(bytes);
    else
        checkMemoryPressureAndUnloadFrameIfNeeded();
```

maybe we shouldn't propagate it upwards and just simply check memory pressure and unload?

Reply:
ok, after carefully thinking about this i have removed the propagating logic. our model is different than RM's.

RM aggregates usage (deltas) in a bottom-up manner, FrameMemoryMonitor's model is self-contained, and the decision should be made at the same level.